File: Python_examples/Maze.py
maze = {
    0: {'north': None,     'east': 1},
    1: {'east': 2,         'south': 4, 'west': 0},
    2: {'south': 5,        'west': 1},
    3: {'east': 4,         'south': 6},
    4: {'north': 1,        'west': 3},
    5: {'north': 2,        'south': 8},
    6: {'north': 3},
    7: {'east': 8},
    8: {'north': 5,        'west': 7}
}

def walk_maze(start, moves):
    pos = start
    for move in moves:
        if move in maze[pos]:
            next_pos = maze[pos][move]
            if next_pos is None:
                print(f"{pos} --{move}--> STOP")
                return
            print(f"{pos} --{move}--> {next_pos}")
            pos = next_pos
        else:
            print(f"{pos} --{move}--> invalid")
            return

walk_maze(0, ['east', 'east', 'south', 'south', 'north', 'north', 'west'])

File: Go/basic_model/Maze.go
package main

import "fmt"

func maze(start int, moves chan string) {
	pos := start
	for m := range moves {
		switch pos {
		case 0:
			if m == "east" {
				pos = 1
			} else if m == "north" {
				fmt.Println("STOP")
				return
			}
		case 1:
			if m == "east" {
				pos = 2
			} else if m == "south" {
				pos = 4
			} else if m == "west" {
				pos = 0
			}
		case 2:
			if m == "south" {
				pos = 5
			} else if m == "west" {
				pos = 1
			}
		case 3:
			if m == "east" {
				pos = 4
			} else if m == "south" {
				pos = 6
			}
		case 4:
			if m == "north" {
				pos = 1
			} else if m == "west" {
				pos = 3
			}
		case 5:
			if m == "north" {
				pos = 2
			} else if m == "south" {
				pos = 8
			}
		case 6:
			if m == "north" {
				pos = 3
			}
		case 7:
			if m == "east" {
				pos = 8
			}
		case 8:
			if m == "north" {
				pos = 5
			} else if m == "west" {
				pos = 7
			}
		}
		fmt.Println("Moved", m, "-> at position", pos)
	}
}

func main() {
	moves := make(chan string)
	go maze(0, moves)
	go func() {
		moves <- "east"
		moves <- "south"
		moves <- "north"
		close(moves)
	}()
	select {}
}
